Update API v1.1 draft content

cwrw opened the pull request:
## Context

The draft v1.1 spec is behind a feature flag to allow the ease of finalising any further comments before sharing with providers.

## Changes proposed in this pull request

Update documentation and add further details to each feature and section.
Please note that some attributes were added or moved under the offer object for consistency.

## Guidance to review

Please add any further comments/ideas that I have missed, draft version can be found here: https://apply-for-te-update-api-jkbuvt.herokuapp.com/api-docs/draft

## Link to Trello card

<!-- http://trello.com/123-example-card -->

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

JR-G commented on app/views/api_docs/vendor_api_docs/reference/draft.html.erb:
```diff
@@ -40,80 +40,101 @@
+
+<p class="govuk-body">To update an interview, the <%= govuk_link_to 'update endpoint', '#post-applications-application_id-interviews-interview_id-update' %> can be invoked with only the fields that require update.</p>
+
+<p class="govuk-body">To cancel an interview, the <%= govuk_link_to 'cancel endpoint', '#post-applications-application_id-interviews-interview_id-cancel' %> can be invoked with the cancellation reason. If an interview is cancelled the <code>cancelled_at</code> and <code>cancellation_reason</code> attributes will be populated</p>
```
sorry!! 😬 nit - full stop missing from here

JR-G reviewed (approved):
Have gone through the changes on the review app – LGTM 👍

mnacos commented on config/vendor_api/v1.1.yml:
```diff
@@ -1179,19 +1176,24 @@ components:
+      - id
       - message
-      - date
+      - updated_at
```
Wondering whether we should publish `created_at` instead of `updated_at`. Notes are special in that we currently don't allow edits. If the content of a note is tweaked via the rails console, we probably still want to show them in `created_at` order.

duncanjbrown replied:
I agree

mnacos replied:
We've added `created_at` to both notes and interviews, but have also kept `updated_at`.

duncanjbrown replied:
👌

duncanjbrown commented:
Could we standardise on "confirm" for the act of making a deferred offer viable? cc @adamsilver https://ukgovernmentdfe.slack.com/archives/CPH8J9G65/p1638790602379500?thread_ts=1638789873.379400&cid=CPH8J9G65

duncanjbrown commented:
Reinstating deferral: "This call only succeeds if the same course, location and study_mode combination is available in the current recruitment cycle." What is the HTTP code and response in this case?

duncanjbrown commented on app/views/api_docs/vendor_api_docs/reference/draft.html.erb:
```diff
@@ -40,80 +40,101 @@
 <p class="govuk-heading-s">Withdrawing applications</p>
 <p class="govuk-body">
-  We now support withdrawing an application at the candidate's request.
+  We now support <%= govuk_link_to 'withdrawing', '#post-applications-application_id-withdraw' %> an application at the candidate's request. If an application is withdrawn at a candidate's request, it will either transition to the <code>declined</code> state if the application is in the <code>offer</code> state, or the <code>withdrawn</code> state if it's in any other state. We currently don't re ...
```
I think this path is a bit ambiguous as `withdraw` offer isn't a separate path. consider explicit `withdraw-offer` (ie alias for reject) and `withdraw-application` paths?

duncanjbrown replied:
Also "withdraw an application or decline an offer" is confusing in the docs, I think — doesn't withdrawal cover both, with declining any current offers as a side effect?

mnacos replied:
It is confusing, we have since tried to improve the description. One reason we steered clear of `withdraw-offer` is that we are not tackling normal offer withdrawal. Instead, this is withdrawing an application at a candidate's request, which results either in a withdrawn application or a declined offer.

duncanjbrown replied:
Yes, exactly! this is withdrawing an application, and the other thing called `withdraw` is withdrawing an offer, so intention was to make it explicit what was being withdrawn in each case

cwrw replied:
`withdraw-offer`: the `reject` endpoint currently withdraws an offer OR rejects an application and we are explicitly hiding the `offer_withdrawn` status. We have moved away from separating those into their own endpoints yet because the work for R4R will not be in v1.1. When we do move towards that, we will be adding that endpoint.

`withdraw` vs `withdraw-application`: if we are to move to separating which we will eventually, any action on `/applications` is implied that it is on the application, so our preference is to keep the action itself as is. We are going to test the waters with providers around this endpoint anyway and with adequate explanation of what the endpoint does we're hoping it'll be clear enough.

duncanjbrown replied:
Great. Makes sense

duncanjbrown commented on config/vendor_api/v1.1.yml:
```diff
@@ -493,7 +493,7 @@ paths:
       summary: Create note
```
This opens a small tin of worms about identity of users in SRSs who are now able to perform actions as though they had accounts on Manage. To be cautious, could we require `email_address`? Or even reuse the metadata field? Or perhaps even for now, descope creation and only allow reads?

mnacos replied:
Plan here is to use the attribution part of MetaData, which is required for POST requests. I think this enforces  the presence of an email address. We've been talking about refactoring notes to have a polymorphic association with authors, supporting both `ProviderUser` and `VendorAPIUser` authors.

Allowing the creation of Notes is probably a good thing, as it would enable more information sharing between people using Manage and people using SRS systems, I think.

duncanjbrown replied:
It is a good thing, but I think the identity q still stands. 2 qs, I guess:

1. what's the point of including a "name" field if we're relying on metadata?
2. as far as I know many providers are using a generic address in the metadata field — does including optional free-text "name" field meaningfully mitigate that?

mnacos replied:
Ah, I see the confusion. When creating a note via the API, vendors are supposed to pass a `CreateNote` object, which only contains a `message` value. Our code reads the metadata, creates/updates a VendorAPIUser and assigns them as the author of the note. When their system retrieves this application later (or immediately, actually, as we respond with the updated application), this contains a `notes:` section consisting of an array of Note objects, each containing a free-text `author` field, derived from the VendorAPIUser. So, `author` is read-only/part of the response.

mnacos replied:
To answer 2, we are not sure what to do about generic addresses in the metadata. The only impact is within Manage, in places like timeline and the activity log, if they are using both Manage and an SRS. And I guess it'd be easy to spot what is going on after looking at a few notes.

duncanjbrown replied:
ah great, makes sense. Thanks!

cwrw replied:
worth mentioning we did raise this with some providers (generic emails etc), and some wanted to take it away

mnacos reviewed (commented):
Looks good. Have added a comment about timestamps for notes.

Please note that `offer_deferred_at` may fit better within the Offer object conceptually but the fields exist within application_choices in the db, so we'll be adding some work to the presenter. But I guess it's a better interface for vendors.

duncanjbrown commented:
Is it weird not to support `GET /applications/{application_id}/interviews`?

duncanjbrown commented:
What's the behaviour of updating an interview when an optional field is present in the record but missed out of the update? does it delete that field? is this really a `PUT`?

duncanjbrown commented:
We'll need to fix the formatting 😅 

<img width="1362" alt="Screenshot 2021-12-07 at 09 56 18" src="https://user-images.githubusercontent.com/642279/145007439-64487086-d9f6-4803-9c38-0d032df006c7.png">

duncanjbrown commented:
Can we take this opportunity to remove `confirm-enrolment`? assuming nobody's using it

cwrw commented:
@JR-G @mnacos to action:
- change URI for reconfirming a deferred offer to `reconfirm-offer`
- swap out reinstate with reconfirm deferred offer
> Reinstating deferral: "This call only succeeds if the same course, location and study_mode combination is available in the current recruitment cycle." What is the HTTP code and response in this case?
- Add 422 with indication of a course not matching in the next year. Also update content around an application moving to pending conditions if they did not meet yet.
- remove "or decline an offer" and just keep the side effect of the status description
- add `created_at` and `updated_at` to notes and interviews
- "To update an interview, the update endpoint can be invoked with only the fields that require update." add more description to indicate that if a field isn't included it won't be overriden

duncanjbrown commented:
> * change URI for reconfirming a deferred offer to `reconfirm-offer`
> * swap out reinstate with reconfirm deferred offer

Is it "reconfirm"? I think Adam would disagree!?

mnacos commented:
> Is it weird not to support `GET /applications/{application_id}/interviews`?

It'd be weird if we had RESTful API but we don't. The plan for these minor releases is to keep returning the full application as part of the response for all types of call, as all are related to an application, to make life easy for vendors/intergrators, who have already written code for parsing applications and would just need to extend it.

Maybe something to revisit for v2.

mnacos commented:
> Can we take this opportunity to remove `confirm-enrolment`? assuming nobody's using it

We talked about it and decided only to remove deprecated fields/endpoints in v2.

mnacos commented:
> What's the behaviour of updating an interview when an optional field is present in the record but missed out of the update? does it delete that field? is this really a `PUT`?

We've added a clarification to the docs that only fields passed along with the request are overwritten and fields already set but not mentioned are preserved. This would be a `PUT` normally, but this is not a RESTful API and we are sticking with GET/POSTs.

mnacos commented:
> Reinstating deferral: "This call only succeeds if the same course, location and study_mode combination is available in the current recruitment cycle." What is the HTTP code and response in this case?

It's a 422 with a json error mentioning the availability of these things in the new cycle. We have updated the docs to make this clearer.

duncanjbrown commented:
> > Is it weird not to support `GET /applications/{application_id}/interviews`?
> 
> It'd be weird if we had RESTful API but we don't. The plan for these minor releases is to keep returning the full application as part of the response for all types of call, as all are related to an application, to make life easy for vendors/intergrators, who have already written code for parsing applications and would just need to extend it.
> 
> Maybe something to revisit for v2.

Great. I'm _very_ here for "this is not a RESTful API" discourse 😄

mnacos commented:
> > > What's the behaviour of updating an interview when an optional field is present in the record but missed out of the update? does it delete that field? is this really a `PUT`?
> > 
> > 
> > We've added a clarification to the docs that only fields passed along with the request are overwritten and fields already set but not mentioned are preserved. This would be a `PUT` normally, but this is not a RESTful API and we are sticking with GET/POSTs.
> 
> Great. Not a PUT by behaviour either

Yes, I guess the behaviour described would be a PATCH, but we don't support that either :-)

cwrw commented:
Thanks @mnacos and @JR-G for addressing all the comments as well as responding to any queries! @duncanjbrown I believe all your comments have been addressed bar the reconfirm which we'll be finalising tomorrow morning, if you can have another look/approve please
